enhancements/etcd/automated-cluster-backups.md

```diff
@@ -0,0 +1,137 @@
+is important to take automated periodical backups to protect the user from unexpected data loss in such scenarios.
+
+## Goals
+
+1.  Automatic periodic backups and automatic pruning
+1.  Make most recent backups readily available
+1.  Eliminate SSH requirement
+
+## Non-Goals
+
+*  Automated restoration is not considered for this enhancement.
+
+## Currently Supported Functionality 
+
+Currently the scripts available support the following functionality:
+### Cluster Backup 
+Takes a snapshot of cluster’s etcd data along with static-pod-resources at the time of the backup. 
+
+### Cluster Restore 
+Restores the etcd data from a backup snapshot. It also restores the static pod resources while deleting all the newer
+revisions.
+
+## Proposal
+
+### A new container is added to etcd static pod to run automated backups. 
+The objectives of the new container are:
+1.  On all etcd static pods, run a subcommand of cluster-etcd-operator to produce automated backups.
+1. On all masters it creates a new backup revision `/etc/kubernetes/static-pod-backups/backup-N`.
+1. On all masters it write `/etc/kubernetes/static-pod-backups/backup-N/backup.env` file containing 4 environmental
+   variables `CREATED`, `OCP_VERSION`, `ETCD_REVISION` and `APISERVER_REVISION`.
+1. On all masters take an etcd snapshot `etcdctl snapshot save
+    /etc/kubernetes/static-pod-backups/backup-N/etcd-data/backup.db`.
+1. On all masters copy all static pod resources  to `/etc/kubernetes/static-pod-backups/backup-N/`.
+1. On all masters symbolically link the directory `/etc/kubernetes/static-pod-backups/latest-backup` with the directory
+   containing the most recent `/etc/kubernetes/static-pod-manifests/backup-N` directory.
+1.  Also be responsible for pruning older backups to keep no more than X number of backups.
+
+###
+
+## User Stories [optional]
+
+### Security
+
+Your clusters backup data is as secure as your cluster. If someone were to root the system they would have direct
+access to all data.
+
+### Availability
+
+Your data is as resilient as your cluster. We make N copies of your data so in the case of failure you dont have to
+worry about your last backup location.
+
+### Recovery Automation
+
+If the cluster were to lose quorum and every master is seeded with data required to restore. Automation of recovery
+tasks becomes easier.
+
+
+## Implementation Plan
+
+1. Create a subcommand to cluster-etcd-operator to run as a container in the etcd static pod.
+2. It runs in an auto-pilot mode without requiring adding any new OpenShift API (for 4.4)
+3. In future releases, cluster-etcd-operator will subsume this functionality with a backup controller (4.6+).
+
+## Implementation Details/Notes/Constraints
+
+*  Add a subcommand to `cluster-etcd-operator`
+*  Change the etcd spec `etcd-pod.yaml` to run the subcommand as a separate container.
+
+## Risks and Mitigations
```

Comment:
Some risks of the sidecar pattern in this context:

* Future changes to the sidecar container spec will require a new etcd pod revision even though the backup mechanism is actually orthogonal to the managed etcd
* Resource requirements of the backup controller are conflated with the requirements for etcd within the pod (what are the possible impacts?)
* Health/liveness checks of the backup controller are conflated with etcd within the pod

Are there any backup controller failure modes which could either take down the etcd pod itself or cause the etcd pod to erroneously report trouble with etcd itself?

Reply:
Another general set of risk related thoughts I wanted to get out of my head...

I think these statements are true:

1. The impact of the platform accidentally killing (or otherwise blowing up) any given etcd container or pod is extremely high
2. The impact of such a problem manifesting against _all etcd pods simultaneously_ is probably catastrophic

If so, then I also think it follows the baseline risk of introducing _any_ additional stuff to the etcd pod is probably pretty high to begin with.

The sidecar approach means adding a lot of new code inside all etcd pods. New code, and code which will have bugs. How confident are we in the boundaries between this sidecar and etcd?

All this is to say, let's pay careful attention to the risks section and try to think of all the ways this could go very wrong.

Reply:
Add performance concerns around snapshot on each node.

Reply:
Added a list of concerns.

Reply:
> If so, then I also think it follows the baseline risk of introducing _any_ additional stuff to the etcd pod is probably pretty high to begin with.

This is not a natural result of those statements.  Simple counter example: point to point network checks.  Your concerns are good, but your conclusion doesn't appear to flow to me.



> The sidecar approach means adding a lot of new code inside all etcd pods. New code, and code which will have bugs. How confident are we in the boundaries between this sidecar and etcd?

We are extremely confident in correct handling of container sized blast radius by a kubelet.  You can seem multiple sidecars, our willingness to add more, and the low risks present by individual failure in the kube-apiserver.

Reply:
Thanks for the thoughts, very helpful